### PR TITLE
Added AWS lambda code for healthchecks

### PR DESCRIPTION
**Pull request by mike-parkhill**

### CODE REVIEW

File: .maintain/monitoring/healthcheck-lambda.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      headers: {
+        'Content-Type': 'application/json',
+        'Content-Length': data.length
+      },
+      timeout: 3000
+    };
+
+    const req = protocol.request(options, res => {
+      if (res.statusCode === 200 || res.statusCode === 400) {
+        res.on('data', d => {
```
Comment (lovesh) [inline .maintain/monitoring/healthcheck-lambda.js:101]: Shouldn't 400 status code be a reject?

File: .maintain/monitoring/healthcheck-lambda.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  },
+  {
+    network: 'Mainnet',
+    url: '{PUT IP/HOSTNAME HERE}',
+    name: 'PoS Full Node',
+    region: 'us-west-1'
+  }
+];
+
+async function parseStatus(status, node) {
```
Comment (lovesh) [inline .maintain/monitoring/healthcheck-lambda.js:21]: This function does more than parsing status so should be appropriately named

**Review by lovesh — CHANGES_REQUESTED** [review]

File: .maintain/monitoring/healthcheck-lambda.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
+const https = require('https');
+const http = require('http');
+const url = require('url');
+const webhookUrl = '{SLACK NOTIFCATION WEBHOOK URL}';
+const port = 9933;
+const apiUrls = [
```
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:6]: the actual lambda checks both mainnet validators. I wasn't sure that the examples here needed to replicate everything in production.

File: .maintain/monitoring/healthcheck-lambda.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    req.end();
+  });
+  return promise;
+}
+
+async function checkAPIStatus(node) {
+  const blockNumberReq = {
+    id: 1,
+    jsonrpc: '2.0',
+    method: 'eth_blockNumber'
```
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:132]: `system_number` returns `Method not found`

```
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32601,
    "message": "Method not found"
  },
  "id": 1
}
```

File: .maintain/monitoring/healthcheck-lambda.js (hunk 1 above)
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:101]: yep - copied without thinking

File: .maintain/monitoring/healthcheck-lambda.js (hunk 2 above)
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:21]: refactored and renamed

File: .maintain/monitoring/healthcheck-lambda.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  if (status.error && status.error.code) {
+    slackBlocks.attachments[0].blocks.push({
+      type: 'section',
+      text: {
+        type: 'mrkdwn',
+        text: `*Code:* ${status.error.code}`,
+      }
+    });
+  }
+  if (status.error && status.error.message) {
```
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:71]: fixed

File: .maintain/monitoring/healthcheck-lambda.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+const http = require('http');
+const url = require('url');
+const webhookUrl = '{SLACK NOTIFCATION WEBHOOK URL}';
+const port = 9933;
+const apiUrls = [
+  {
+    network: 'Mainnet',
+    url: '{PUT IP/HOSTNAME HERE}',
+    name: 'PoS Validator #1',
+    region: 'us-east-2'
```
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js:11]: done

File: .maintain/monitoring/README.md [hunk 7]
```diff
@@ -11,3 +11,7 @@ Two files `prometheus.yaml` and `alerting-rules.yaml` are used for prometheus an
 ## Setup guide
 
 The good people at robonomics have created a nice guide to get you started: https://github.com/hubobubo/robonomics/wiki/Robonomics-(XRT)-metrics-using-Prometheus-and-Grafana - you can follow this and import the dashboard JSON here or their panel from Grafana.
+
+## Healthcheck Lambda
+There is an AWS Lambda running as a healthcheck for all PoS nodes (full and validator).
+The code for it is in `./healthcheck-lambda.js`.
```
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/README.md:17]: done

File: .maintain/monitoring/healthcheck-lambda.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply .maintain/monitoring/healthcheck-lambda.js:6]: No, you don't need to. I was just checking what was happening in the actual lambda

File: .maintain/monitoring/healthcheck-lambda.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply .maintain/monitoring/healthcheck-lambda.js:132]: My bad, can you please check `system_syncState`

**Comment by lovesh** [comment]
Have you pushed the commit with changes?

**Comment by mike-parkhill** [comment]
> 
> 
> Have you pushed the commit with changes?

No, I realized that the full nodes and Slack were unreachable again when I tried testing it out. I thought I'd fixed that last night. But I can probably push up my updates while I fiddle with AWS.

File: .maintain/monitoring/healthcheck-lambda.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
+const https = require('https');
+const http = require('http');
+const url = require('url');
+const webhookUrl = process.env.SLACK_NOTIFCATION_WEBHOOK_URL;
```
Comment (mike-parkhill) [inline .maintain/monitoring/healthcheck-lambda.js:4]: I switched this and the node url JSON config to be env vars so we don't have to fiddle with the code when pushing out updates.

File: .maintain/monitoring/healthcheck-lambda.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+const port = 9933;
+
+async function hasErrorStatus(status) {
+  if (status.result && status.result.startsWith('0x')) return false;
```
Comment (lovesh) [inline .maintain/monitoring/healthcheck-lambda.js]: `return !(status.result && status.result.startsWith('0x'))` as function body is sufficient

**Review by lovesh — APPROVED** [review]

File: .maintain/monitoring/healthcheck-lambda.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mike-parkhill) [reply .maintain/monitoring/healthcheck-lambda.js]: That worked.

**Pull request merged**